src/lib/os-release.c

```diff
@@ -0,0 +1,235 @@
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+
+/**
+ * Statically defined fields for correctness of implementation
+ */
+static const char *os_release_fields[] = {
+            [OS_RELEASE_NAME] = "NAME",
+            [OS_RELEASE_VERSION] = "VERSION",
+            [OS_RELEASE_ID] = "ID",
+            [OS_RELEASE_VERSION_ID] = "VERSION_ID",
+            [OS_RELEASE_PRETTY_NAME] = "PRETTY_NAME",
+            [OS_RELEASE_ANSI_COLOR] = "ANSI_COLOR",
+            [OS_RELEASE_HOME_URL] = "HOME_URL",
+            [OS_RELEASE_SUPPORT_URL] = "SUPPORT_URL",
+            [OS_RELEASE_BUG_REPORT_URL] = "BUG_REPORT_URL",
+};
+
+/**
+ * Return a sane fallback key if one isn't provided in the config.
+ */
+static const char *cbm_os_release_fallback_value(CbmOsReleaseKey key)
+{
+        switch (key) {
+        case OS_RELEASE_NAME:
+                return "generic-linux-os";
+        case OS_RELEASE_PRETTY_NAME:
+                return "generic-linux-os";
+        case OS_RELEASE_ID:
+        case OS_RELEASE_VERSION:
+        case OS_RELEASE_VERSION_ID:
+                return "1";
+        default:
+                return "";
+        }
+}
+
+/**
+ * Do the actual hard work of parsing the os-release file.
+ */
+static bool cbm_os_release_parse(CbmOsRelease *self, const char *path)
+{
+        autofree(FILE) *f = NULL;
+        size_t sn;
+        ssize_t r = 0;
+        char *buf = NULL;
+        bool ret = true;
+
+        f = fopen(path, "r");
+        if (!f) {
+                if (errno != ENOENT) {
+                        LOG_ERROR("Unable to open %s: %s", path, strerror(errno));
+                }
+                return false;
+        }
+
+        while ((r = getline(&buf, &sn, f)) > 0) {
+                ssize_t cur = 0;
+                size_t val_len = 0;
+                size_t incr = 0;
+                char *key = NULL;
+                autofree(char) *value = NULL;
+
+                /* Strip newlines */
+                if (r >= 1 && buf[r - 1] == '\n') {
+                        buf[r - 1] = '\0';
+                        --r;
+                }
+
+                char *l = buf;
+                char *c = NULL;
+
+                /* Skip empty lines */
+                if ((r == 1 && buf[0] == '\n') || r < 1) {
+                        goto next_line;
+                }
+
+                /* Skip the starting whitespace */
+                while (isspace(*l)) {
+                        ++l;
+                        if (!*l) {
+                                break;
+                        }
+                }
+
+                /* Skip a comment */
+                if (l[0] == '#') {
+                        goto next_line;
+                }
+
+                cur = (l - buf);
+                r -= cur;
+
+                /* Look for assignment */
+                c = memchr(l, '=', (size_t)r);
+                if (!c) {
+                        goto next_line;
+                }
+
+                /* Skip empty value */
```

Comment:
So it seems stupid but should probably have an empty key check too.

That said the current code is safe since strndup will allocate a `\0` for the 0 length key.

As an aside I'm wondering about whitespace stripping before and after = and at the end of the value.